# Profitable app

- Our company builds apps that are free to download and nstall, and our main source of revenue consists of in-app ads. This means our revenue for any given app is mostly influenced by the number of users who use our app.
- Analyzing data to help our developers understand what type of apps are likely to attract more users.

In [1]:
from csv import reader

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))                    

In [3]:
def open_file(path):
    with open(path,encoding='UTF-8') as f:
        read_f=reader(f)
        data=list(read_f)
        
        return data

In [4]:
ios=open_file('AppleStore.csv')
android=open_file('googleplaystore.csv')
ios_header=ios[0]
ios=ios[1:]
android_header=android[0]
android=android[1:]

In [5]:
print(android_header)
print('\n')
explore_data(android,0,2,True)
print('\n')
print(ios_header)
print('\n')
explore_data(ios,0,2,True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558

## Data to work with:
- Android: installs, genres?, category, price
- IoS: There are not so many columns to help. Price, rating_count_tot, prime_genre


In [6]:
print(android_header)
print('\n')
print(android[10472])
del android[10472]

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


The play store database contains duplicate items

In [7]:
for app in android:
    name=app[0]
    if name=='Facebook'or name=='Twitter':
        print(app)

['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Twitter', 'NEWS_AND_MAGAZINES', '4.3', '11667403', 'Varies with device', '500,000,000+', 'Free', '0', 'Mature 17+', 'News & Magazines', 'August 6, 2018', 'Varies with device', 'Varies with device']
['Twitter', 'NEWS_AND_MAGAZINES', '4.3', '11667403', 'Varies with device', '500,000,000+', 'Free', '0', 'Mature 17+', 'News & Magazines', 'August 6, 2018', 'Varies with device', 'Varies with device']
['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Twitter', 'NEWS_AND_MAGAZINES', '4.3', '11657972', 'Varies with device', '500,000,000+', 'Free', '0', 'Mature 17+', 'News & Magazines', 'July 30, 2018', 'Varies with device', 'Varies with device']


In [8]:
unique_apps=[]
duplicate_apps=[]
for app in android:
    name=app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print("The number of duplicate apps",len(duplicate_apps))
print("Some examples of duplicate apps:\n",duplicate_apps[:4])

The number of duplicate apps 1181
Some examples of duplicate apps:
 ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings']


I won't remove duplicates randomly, I will remove them based on the review count, II will keep the apps with the highest number of ratings because this means that app version is the mmost up-to-date. I can't on anything else (on the current version I can't).

In [9]:
name_and_max_reviews={}
for app in android:
    rating_counts=int(app[3])
    name=app[0]
    
    # slack comes in with 50000 reviews
    # it's not in the dictionary so the dict gets updated
    # slack comes again with 51000 reviews the if statement evqluetas to true
    # so the dict gets updated again
    # slack comes in the third time but now with only 50500 reviews both 
    # statements evaluate to false so the dict doesnt get updated. so the name
    # gets thrown away
    if name in name_and_max_reviews and name_and_max_reviews[name]<rating_counts:
        name_and_max_reviews[name]=rating_counts
        
    elif name not in name_and_max_reviews:
        name_and_max_reviews[name]=rating_counts
len(name_and_max_reviews)


9659

In [10]:
android_clean=[]
already_added=[]
for app in android:
    name=app[0]
    reviews=int(app[3])
    
    if reviews==name_and_max_reviews[name] and name not in already_added:
        android_clean.append(app)
        already_added.append(name)
len(android_clean)

9659

In [11]:
def char_check(string)->bool:
    cnt=0
    for c in string:
        cnt+=ord(c)>127
#         print(c,ord(c))
        if cnt>3:            
            return False
    return True
# print(char_check('Instachat 😜😜😜'))
# cnt=0
# print(cnt+True+True)

In [12]:
clean_ios_data=[]
for app in ios:
    if char_check(app[1]):
        clean_ios_data.append(app)
len(clean_ios_data)        

6183

In [13]:
clean_android_data=[]
for app in android_clean:
    if char_check(app[0]):
        clean_android_data.append(app)
len(clean_android_data)

9614

In [14]:
(len(ios)-len(clean_ios_data))-(len(android_clean)-len(clean_android_data))

969

We can see that the android data set has far less non english apps (969
less). Even though the android data set contains more items than the ios.

In [15]:
relevant_android_data=[]
relevant_ios_data=[]
for app in clean_android_data:
    if app[7]=='0':
        relevant_android_data.append(app)
for app in clean_ios_data:
    if float(app[4])==0:
        relevant_ios_data.append(app)
print(len(relevant_ios_data))
print(len(relevant_android_data))

3222
8864


We can see that the app store contains a lot more non-free apps, than the android

In [16]:
print(ios_header)
print('\n')
unique_apps_ios=[]
duplicate_apps_ios=[]
for app in relevant_ios_data:
    if app[1] in unique_apps_ios:
        duplicate_apps_ios.append(app[1])
        print(app)
        print('\n')
    else:
        unique_apps_ios.append(app[1])
print('\n')
print(len(duplicate_apps_ios))
print(len(unique_apps_ios))


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['1178454060', 'Mannequin Challenge', '59572224', 'USD', '0.0', '105', '58', '4.0', '4.5', '1.0.1', '4+', 'Games', '38', '5', '1', '1']


['1089824278', 'VR Roller Coaster', '240964608', 'USD', '0.0', '67', '44', '3.5', '4.0', '0.81', '4+', 'Games', '38', '0', '1', '1']




2
3220


We can see that the relevant apps ios contains only 2 duplicate apps, so we aren't going to go through the same process for the sake of just 2 apps. (maybe if i wrote a function. Also these 2 apps have ~100 reviews).

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

1. Build a minimal Android version of the app, and add it to Google Play.
2. If the app has a good response from users, we develop it further.
3. If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.


In [50]:
def freq_table(lst,idx):
    freq_table={}
    for app in lst:
        genre=app[idx]
        if genre in freq_table:
            freq_table[genre]+=1
        else:
            freq_table[genre]=1
    return freq_table

In [55]:
def show_percent(freq_table)->dict:

    d={}
    
    total_number_of_apps=0
    for i in freq_table:
        total_number_of_apps+=freq_table[i]
    for i in freq_table:
        d[i]=f'{float(freq_table[i]/total_number_of_apps*100):.2f}'
        
    return d

In [61]:
freq_table_android_cat=freq_table(relevant_android_data,1)
freq_table_android=freq_table(relevant_android_data,-4)
freq_table_ios=freq_table(relevant_ios_data,-5)
ios_percs=show_percent(freq_table_ios)
android_percs=show_percent(freq_table_android)
android_cat_percs=show_percent(freq_table_android_cat)

In [93]:
def sorting(dic)->dict:
    d={}
    
    lst=sorted(dic,key=lambda x:(-float(dic[x]),x))
    percs=0
    for i in lst:
        percentage=float(dic[i])
        if percentage<0.17:
            percs+=percentage
        else:
            d[i]=dic[i]
#     if percs>0:
#         d['Other']=f"{float(percs):.2f}"
    return d
sorted_ios_percentages_genres=(sorting(ios_percs))
sorted_android_percentages_genres=(sorting(android_percs))
sorted_android_percentages_cats=(sorting(android_cat_percs))

In [42]:
# weather = {'Reykjavik': {'temp':60, 'condition': 'rainy'},
# 'Buenos Aires': {'temp': 55, 'condition': 'cloudy'},
# 'Cairo': {'temp': 96, 'condition': 'sunny'},
# 'Berlin': {'temp': 89, 'condition': 'sunny'},
# 'Caloocan': {'temp': 78, 'condition': 'sunny'}}
# sorted(weather, key=lambda w: (w, weather[w]['temp']))

In [43]:
print(sorted_android_percentages_genres)
print('\n')
print(sorted_android_percentages_cats)
print('\n')
print(sorted_ios_percentages_genres)

{'Tools': '8.45', 'Entertainment': '6.07', 'Education': '5.35', 'Business': '4.59', 'Lifestyle': '3.89', 'Productivity': '3.89', 'Finance': '3.70', 'Medical': '3.53', 'Sports': '3.46', 'Personalization': '3.32', 'Communication': '3.24', 'Action': '3.10', 'Health & Fitness': '3.08', 'Photography': '2.94', 'News & Magazines': '2.80', 'Social': '2.66', 'Travel & Local': '2.32', 'Shopping': '2.25', 'Books & Reference': '2.14', 'Simulation': '2.04', 'Dating': '1.86', 'Arcade': '1.85', 'Video Players & Editors': '1.77', 'Casual': '1.76', 'Maps & Navigation': '1.40', 'Food & Drink': '1.24', 'Puzzle': '1.13', 'Racing': '0.99', 'Libraries & Demo': '0.94', 'Role Playing': '0.94', 'Auto & Vehicles': '0.93', 'Strategy': '0.91', 'House & Home': '0.82', 'Weather': '0.80', 'Events': '0.71', 'Adventure': '0.68', 'Comics': '0.61', 'Art & Design': '0.60', 'Beauty': '0.60', 'Parenting': '0.50', 'Card': '0.45', 'Casino': '0.43', 'Trivia': '0.42', 'Educational;Education': '0.39', 'Board': '0.38', 'Educatio

## App store

- Games is the most common genre, and Catalogs is the runner up
- I can see that entartaninment has a huge share in the market. Apps that are designed for practical purposes are way less. But this doesn't mean that games have the highest demand.
- Sadly I can not recommend an app profile based on my findings so far.

## Google play store

- For the sake of the bigger picture I am going to analyze only the finidings on the categories column
- The most common genre is Family which contains just games for kids, but here I can see more practical purpose apps than in the app store.
- Sadly I can not recommend an app profile based on my findings so far.

In [62]:
#  1 5 android_header
# -5 5 ios_header

In [63]:
tot_ratings_by_genre_ios={}
for key in freq_table_ios:
    for app in relevant_ios_data:
        if key==app[-5]:
            if key in tot_ratings_by_genre_ios:
                tot_ratings_by_genre_ios[key]+=int(app[5])
            else:
                tot_ratings_by_genre_ios[key]=int(app[5])
            
    tot_ratings_by_genre_ios[key]=int(tot_ratings_by_genre_ios[key]/freq_table_ios[key])

In [64]:
sorting(tot_ratings_by_genre_ios)

{'Navigation': 86090,
 'Reference': 74942,
 'Social Networking': 71548,
 'Music': 57326,
 'Weather': 52279,
 'Book': 39758,
 'Food & Drink': 33333,
 'Finance': 31467,
 'Photo & Video': 28441,
 'Travel': 28243,
 'Shopping': 26919,
 'Health & Fitness': 23298,
 'Sports': 23008,
 'Games': 22788,
 'News': 21248,
 'Productivity': 21028,
 'Utilities': 18684,
 'Lifestyle': 16485,
 'Entertainment': 14029,
 'Business': 7491,
 'Education': 7003,
 'Catalogs': 4004,
 'Medical': 612}

I don't even know why the ios is the benchmark, but every category has a a few apps that excel in reviews. It would be nicer to view the first quartile or the median of the reviews, I'm not gonna do that, for now. Based on the findings so far I think the games category seem to be a promising opportunity to make an app in, because of the amount of games and their avg reviews. But et's see this avg multiplied by the percentage.

In [94]:
tot_ratings_by_genre_ios_percs_multi_reviews={}
for key in sorted_ios_percentages_genres:
    tot_ratings_by_genre_ios_percs_multi_reviews[key]=tot_ratings_by_genre_ios[key]*float(sorted_ios_percentages_genres[key])/100
tot_ratings_by_genre_ios_percs_multi_reviews

{'Games': 13253.500799999998,
 'Entertainment': 1105.4852,
 'Photo & Video': 1413.5176999999999,
 'Education': 256.3098,
 'Social Networking': 2353.9292,
 'Shopping': 702.5858999999999,
 'Utilities': 468.9684,
 'Sports': 492.37120000000004,
 'Music': 1175.183,
 'Health & Fitness': 470.6196,
 'Productivity': 365.8872,
 'Lifestyle': 260.463,
 'News': 282.5984,
 'Travel': 350.2132,
 'Finance': 352.4304,
 'Weather': 454.82730000000004,
 'Food & Drink': 269.99730000000005,
 'Reference': 419.6752,
 'Business': 39.7023,
 'Book': 170.9594,
 'Medical': 1.1628,
 'Navigation': 163.571}

We can see by that that games excel in reviews*distribution

In [87]:
tot_ratings_by_genre_android={}
for key in freq_table_android_cat:
    
    for app in relevant_android_data:
        
        if key==app[1]:
            installs=app[5].replace(',','')
            installs=int(installs.replace('+',''))
            if key in tot_ratings_by_genre_android:
                
                tot_ratings_by_genre_android[key]+=installs
                
            else:
                tot_ratings_by_genre_android[key]=installs
    
    tot_ratings_by_genre_android[key]=int(tot_ratings_by_genre_android[key]/freq_table_android_cat[key])

In [88]:
sorting(tot_ratings_by_genre_android)

{'COMMUNICATION': 38456119,
 'VIDEO_PLAYERS': 24727872,
 'SOCIAL': 23253652,
 'PHOTOGRAPHY': 17840110,
 'PRODUCTIVITY': 16787331,
 'GAME': 15588015,
 'TRAVEL_AND_LOCAL': 13984077,
 'ENTERTAINMENT': 11640705,
 'TOOLS': 10801391,
 'NEWS_AND_MAGAZINES': 9549178,
 'BOOKS_AND_REFERENCE': 8767811,
 'SHOPPING': 7036877,
 'PERSONALIZATION': 5201482,
 'WEATHER': 5074486,
 'HEALTH_AND_FITNESS': 4188821,
 'MAPS_AND_NAVIGATION': 4056941,
 'FAMILY': 3695641,
 'SPORTS': 3638640,
 'ART_AND_DESIGN': 1986335,
 'FOOD_AND_DRINK': 1924897,
 'EDUCATION': 1833495,
 'BUSINESS': 1712290,
 'LIFESTYLE': 1437816,
 'FINANCE': 1387692,
 'HOUSE_AND_HOME': 1331540,
 'DATING': 854028,
 'COMICS': 817657,
 'AUTO_AND_VEHICLES': 647317,
 'LIBRARIES_AND_DEMO': 638503,
 'PARENTING': 542603,
 'BEAUTY': 513151,
 'EVENTS': 253542,
 'MEDICAL': 120550}

Based on my findings I can highly recommend games because they alone have 15.6 mill installs on average if we take the percentages and of the whole bunchand multiply it with the values of the upper dict, we get this: however based on the first quartile or the median should we judge.

In [89]:
tot_ratings_by_genre_android_percentages_multi_installs={}
for key in tot_ratings_by_genre_android:
    tot_ratings_by_genre_android_percentages_multi_installs[key]=tot_ratings_by_genre_android[key]*float(sorted_android_percentages_cats[key])/100

In [90]:
sorting(tot_ratings_by_genre_android_percentages_multi_installs)

{'GAME': 1515155.0580000002,
 'COMMUNICATION': 1245978.2556,
 'TOOLS': 913797.6786000001,
 'FAMILY': 698845.7131,
 'PRODUCTIVITY': 653027.1759,
 'SOCIAL': 618547.1432,
 'PHOTOGRAPHY': 524499.2339999999,
 'VIDEO_PLAYERS': 442628.90880000003,
 'TRAVEL_AND_LOCAL': 327227.4018,
 'NEWS_AND_MAGAZINES': 267376.984,
 'BOOKS_AND_REFERENCE': 187631.15540000002,
 'PERSONALIZATION': 172689.20239999998,
 'SHOPPING': 158329.7325,
 'HEALTH_AND_FITNESS': 129015.6868,
 'SPORTS': 123713.76,
 'ENTERTAINMENT': 111750.76799999998,
 'BUSINESS': 78594.11099999999,
 'MAPS_AND_NAVIGATION': 56797.17399999999,
 'LIFESTYLE': 56074.82399999999,
 'FINANCE': 51344.60400000001,
 'WEATHER': 40595.888000000006,
 'FOOD_AND_DRINK': 23868.7228,
 'EDUCATION': 21268.541999999998,
 'DATING': 15884.9208,
 'ART_AND_DESIGN': 12712.544000000002,
 'HOUSE_AND_HOME': 10918.628,
 'AUTO_AND_VEHICLES': 6020.048100000001,
 'LIBRARIES_AND_DEMO': 6001.928199999999,
 'COMICS': 5069.4734,
 'MEDICAL': 4255.415,
 'PARENTING': 3526.9195,
 'BE